miri-script/src/commands.rs

```diff
@@ -0,0 +1,711 @@
+use itertools::chain;
+use std::collections::BTreeMap;
+
+use std::io::Result;
+use std::path::{Path, PathBuf};
+
+use duct::cmd;
+use path_macro::path;
+use walkdir::WalkDir;
+
+use crate::arg::Subcommands;
+
+const JOSH_FILTER: &str =
+    ":rev(75dd959a3a40eb5b4574f8d2e23aa6efbeb33573:prefix=src/tools/miri):/src/tools/miri";
+
+fn detect_miri_dir() -> std::io::Result<PathBuf> {
+    const MIRI_SCRIPT_ROOT_DIR: &str = env!("CARGO_MANIFEST_DIR");
+    Ok(PathBuf::from(MIRI_SCRIPT_ROOT_DIR).canonicalize()?.parent().unwrap().into())
+}
+
+fn get_current_toolchain(miri_dir: &Path) -> std::io::Result<String> {
+    let stdout = cmd!("rustup", "show", "active-toolchain").dir(miri_dir).read()?;
+    Ok(stdout.split_whitespace().next().expect("Could not obtain active Rust toolchain").into())
+}
+
+fn apply_env_variables<'a>(
+    cmd: &'a mut duct::Expression,
+    variables: &[(String, String)],
+) -> &'a mut duct::Expression {
+    for (key, value) in variables {
+        *cmd = cmd.env(key, value);
+    }
+    cmd
+}
+#[derive(Clone, Debug)]
+pub(super) struct MiriRunner<'a> {
+    /// miri_dir is based on build directory of miri-script.
```

Comment:
This doesn't tell us what directory this is.
```suggestion
    /// miri_dir is the root of the miri repository checkout we are working in.
```